cpp/src/io/parquet/experimental/hybrid_scan_impl.hpp

```diff
@@ -0,0 +1,507 @@
+   */
+  void create_global_chunk_info(parquet_reader_options const& options);
+
+  /**
+   * @brief Computes all of the passes we will perform over the file.
+   */
+  void compute_input_passes();
+
+  /**
+   * @brief Given a set of pages that have had their sizes computed by nesting level and
+   * a limit on total read size, generate a set of {skip_rows, num_rows} pairs representing
+   * a set of reads that will generate output columns of total size <= `chunk_read_limit` bytes.
+   */
+  void compute_output_chunks_for_subpass();
+
+  /**
+   * @brief Check if there is more work to be done.
+   */
+  [[nodiscard]] bool has_more_work() const;
+
+  /**
+   * @brief Read a chunk of data and return an output table.
+   *
+   * This function is called internally and expects all preprocessing steps have already been done.
+   *
+   * @tparam RowMaskView View type of the row mask column
+   * @param[in] read_mode Read mode indicating if we are reading filter or payload columns
+   * @param[in,out] row_mask Boolean column indicating which rows need to be read after page-pruning
+   *                         for filter columns, or after materialize step for payload columns
+   * @return The output table along with columns' metadata
+   */
+  template <typename RowMaskView>
+  table_with_metadata read_chunk_internal(read_mode read_mode, RowMaskView row_mask);
+
+  /**
+   * @brief Check if this is the first output chunk
+   *
+   * @return True if this is the first output chunk
+   */
+  [[nodiscard]] bool is_first_output_chunk() const;
+
+ private:
+  using named_to_reference_converter = parquet::detail::named_to_reference_converter;
+  using input_column_info            = parquet::detail::input_column_info;
+  using inline_column_buffer         = io::detail::inline_column_buffer;
+  using reader_column_schema         = io::reader_column_schema;
+  using file_intermediate_data       = parquet::detail::file_intermediate_data;
+  using pass_intermediate_data       = parquet::detail::pass_intermediate_data;
+
+  rmm::cuda_stream_view _stream;
+  rmm::device_async_resource_ref _mr{cudf::get_current_device_resource_ref()};
+
+  std::unique_ptr<aggregate_reader_metadata> _metadata;
+
+  // name to reference converter to extract AST output filter
+  named_to_reference_converter _expr_conv{std::nullopt, table_metadata{}};
+
+  // input columns to be processed
+  std::vector<input_column_info> _input_columns;
+  // Buffers for generating output columns
+  std::vector<inline_column_buffer> _output_buffers;
+  // Buffers copied from `_output_buffers` after construction for reuse
+  std::vector<inline_column_buffer> _output_buffers_template;
+  // _output_buffers associated schema indices
+  std::vector<int> _output_column_schemas;
+
+  // _output_buffers associated metadata
+  std::unique_ptr<table_metadata> _output_metadata;
+
+  std::optional<std::vector<std::string>> _filter_columns_names;
+
+  bool _strings_to_categorical = false;
```

Comment:
inconsistent style with initializations below

Reply:
I am blaming `reader_impl.hpp` for this one 😆